### PR TITLE
[SPARK-23149][SQL] polish ColumnarBatch

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Several cleanups in `ColumnarBatch`
* remove `schema`. The `ColumnVector`s inside `ColumnarBatch` already have the data type information, we don't need this `schema`.
* remove `capacity`. `ColumnarBatch` is just a wrapper of `ColumnVector`s, not builders, it doesn't need a capacity property.
* remove `DEFAULT_BATCH_SIZE`. As a wrapper, `ColumnarBatch` can't decide the batch size, it should be decided by the reader, e.g. parquet reader, orc reader, cached table reader. The default batch size should also be defined by the reader.

## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @hvanhovell @gatorsmile @viirya @kiszk

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java [hunk 1]
```diff
@@ -192,7 +192,7 @@ private static void appendValue(WritableColumnVector dst, DataType t, Row src, i
    */
   public static ColumnarBatch toBatch(
       StructType schema, MemoryMode memMode, Iterator<Row> row) {
-    int capacity = ColumnarBatch.DEFAULT_BATCH_SIZE;
+    int capacity = 4 * 1024;
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java]: Ur, is it okay? It seems that we need a constant variable like before.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 2]
```diff
@@ -50,6 +50,9 @@
  * TODO: make this always return ColumnarBatches.
  */
 public class VectorizedParquetRecordReader extends SpecificParquetRecordReaderBase<Object> {
+  // TODO: make this configurable.
+  private static final int CAPACITY = 4 * 1024;
```
Comment (jiangxb1987) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Do we want to have a consistent default capacity for `OrcColumnarBatchReader` and `VectorizedParquetRecordReader`, or have them set separately?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java [hunk 3]
```diff
@@ -98,22 +88,22 @@ public OrcColumnarBatchReader(boolean useOffHeap, boolean copyToSpark) {
 
 
   @Override
-  public Void getCurrentKey() throws IOException, InterruptedException {
+  public Void getCurrentKey() {
```
Comment (jiangxb1987) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: Do we also want to apply similar changes to `SpecificParquetRecordReaderBase` ？

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala [hunk 4]
```diff
@@ -90,8 +90,9 @@ case class ArrowEvalPythonExec(udfs: Seq[PythonUDF], output: Seq[Attribute], chi
 
       private var currentIter = if (columnarBatchIter.hasNext) {
         val batch = columnarBatchIter.next()
-        assert(schemaOut.equals(batch.schema),
-          s"Invalid schema from pandas_udf: expected $schemaOut, got ${batch.schema}")
+        val actualDataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType())
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: Emmm... Looks good to have it inlined for now, in the future if we are to fetch the schema elsewhere, we may move this to `ColumnVectorUtils`.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java]: places using `ColumnarBatch` should decide the default size themselves.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: we should set them separately, places using `ColumnarBatch` should decide the default size themselves.

Comment (jiangxb1987) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Then, what should we use for capacity in `ColumnVectorUtils.toBatch()` ?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java [hunk 5]
```diff
@@ -49,18 +49,8 @@
...
-   * columnar batch, because their default batch sizes are different like the following:
-   *
-   * - ORC's VectorizedRowBatch.DEFAULT_SIZE = 1024
-   * - Spark's ColumnarBatch.DEFAULT_BATCH_SIZE = 4 * 1024
-   */
-  private static final int DEFAULT_SIZE = 4 * 1024;
-
-  // ORC File Reader
-  private Reader reader;
+  // TODO: make this configurable.
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java]: The comment is not valid now. Spark `ColumnarBatch` doesn't have a default size now, the reader just need to decide the capacity itself.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: I also pick `4 * 1024`, but that method would be removed eventually, it's not used anywhere, just an example for writing rows to batches.

Comment (jiangxb1987) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: If this is just an example, how about move it to the ColumnarBatchSuite?

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: but it's not a test, it's an example about how to implement a `ColumnarBatch` producer. We would remove it when we publish `WritableColumnVector` and have a stable API for writing `ColumnarBatch`.

**Comment by jiangxb1987** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnarBatch.java [hunk 6]
```diff
@@ -96,16 +90,6 @@ public void setNumRows(int numRows) {
...
 
-  /**
-   * Returns the schema that makes up this batch.
-   */
-  public StructType schema() { return schema; }
-
-  /**
-   * Returns the max capacity (in number of rows) for this batch.
-   */
-  public int capacity() { return capacity; }
```
Comment (kiszk) [inline sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: I agree to remove these fields `schema` and `capacity` from `ColumnarBatch`.
Is it better to prepare APIs to get `schema` and `capacity` from a set of `ColumnVector`s?

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: For `ColumnarBatch` consumers, they don't care about `capacity`, but only `numRows`, 
`capacity` is only needed by column vector builders. Also they don't care about schema/field names, but only the data type of each column.

**Pull request closed without merging**